resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java

```diff
@@ -56,13 +164,106 @@ public void close() throws IOException
+         queue(op);
+         return op.future;
+      }
+
+      @Override
+      public CompletionStage<Void> asyncWrite(byte[] bytes, int offset, int length)
+      {
+         AsyncOperation op = new WriteOperation(this, bytes, offset, length);
+         queue(op);
+         return op.future;
+      }
+
+      private void queue(AsyncOperation op)
+      {
+         // fetch it from the context directly to avoid having to restore the context just in case we're invoked on a context-less thread
+         HttpRequest resteasyRequest = (HttpRequest) contextDataMap.get(HttpRequest.class);
+         if(request.isAsyncStarted() && !resteasyRequest.getAsyncContext().isOnInitialRequest()) {
+            synchronized(this) {
+               ServletOutputStream os;
+               try
+               {
+                  os = response.getOutputStream();
+               } catch (IOException e)
+               {
+                  // return a failed future, do not queue it
+                  op.future.completeExceptionally(e);
+                  return;
+               }
+               if(!asyncRegistered) {
+                  // start the queue
+                  asyncRegistered = true;
+                  // make sure we have something ready to be executed
+                  addToQueue(op);
+                  os.setWriteListener(this);
+               } else if(os.isReady()) {
+                  // it's possible that we startAsync and queue, then queue another event and the stream becomes ready before
+                  // onWritePossible is called, which means we need to flush the queue here to guarantee ordering if that happens
+                  flushQueue(os);
+                  // we're not allowed to queue work if the output is ready
+                  lastAsyncOperation = op;
+                  op.work(os);
+               } else {
+                  // just queue
+                  addToQueue(op);
+               }
+            }
+         } else {
+            op.work(null);
+         }
+      }
+
+      private void flushQueue(ServletOutputStream sos)
+      {
+         if(lastAsyncOperation != null) {
+            lastAsyncOperation.future.complete(null);
+            lastAsyncOperation = null;
+         }
+
+         while(!asyncQueue.isEmpty() && sos.isReady()) {
+            lastAsyncOperation = asyncQueue.poll();
+            lastAsyncOperation.work(sos);
+         }
+      }
+
+      private synchronized void addToQueue(AsyncOperation op)
+      {
+         if(asyncQueue == null) {
+            asyncQueue = new ConcurrentLinkedQueue<>();
+         }
+         asyncQueue.add(op);
+      }
+
+      @Override
+      public synchronized void onWritePossible() throws IOException
+      {
+         flushQueue(response.getOutputStream());
+      }
+
+      @Override
+      public synchronized void onError(Throwable t)
+      {
+         if(lastAsyncOperation != null) {
+            lastAsyncOperation.future.completeExceptionally(t);
```

Comment:
Do where need to complete the full queue here?

Reply:
Well that's the limit of my understanding of this brain-dead servlet async IO API. I guessed the error was for a random operation, so I took the last one. No idea if the error applies to all queued async ops. Do you know?

Reply:
Once there is an error everything is broken, you are not going to get any more notifications as writing is done.

Reply:
OK done.